pkgs/servers/http/nginx/0001-Support-mixed-addr-configuration-for-PROXY-protocol-.patch

```diff
@@ -0,0 +1,58 @@
+From a48b3cc3c0ee344673a9f508f7b4a1b754a389cd Mon Sep 17 00:00:00 2001
+From: Raito Bezarius <masterancpp@gmail.com>
+Date: Mon, 30 Jan 2023 04:06:16 +0100
+Subject: [PATCH] Support mixed addr configuration for PROXY protocol in HTTP
+ realip module
+
+This ensures that under `real_ip_header proxy_protocol`, if a request is
+received on a `listen` block which do not contain `proxy_protocol`, it
+is not rejected by NGINX.
```

Comment:
I didn't think this through yet but how dangerous would this be with the following headers recommended by upstreams doc?
```
set_real_ip_from ${somePrivateIpv4};
set_real_ip_from ${somePrivateIpv6};

real_ip_header proxy_protocol;

proxy_set_header X-Real-IP       $proxy_protocol_addr;
proxy_set_header X-Forwarded-For $proxy_protocol_addr;
```

Reply:
Let's try it! I will add something to the tests with that and let's see if we can create a situation where we spoof IPs.

Reply:
Thinking out loud and trying some stuff:

- Without this patch, if you listen with proxy protocol, you can still be spoofed if you don't firewall properly the entering packets, and you trust blindly `$proxy_protocol_addr`
- `set_real_ip_from` is only a simple filtering control solution which aims to replace automatically `$remote_addr` and `$remote_port` out of the box based on `real_ip_header` directive.
- Upstream advises to use `set_real_ip_from` **OR** `proxy_set_header X-Real-IP ...`, etc. But upstream is a bit confusing in how they advise to do it.

In your scenario, I indeed can control `X-Real-IP` and `X-Forwarded-For` even under `set_real_ip_from` which is not here to prevent that.

My patch BTW under no proxy protocol does:
`            ngx_http_set_ctx(r, NULL, ngx_http_realip_module);`

which clears all the variables sets by NGINX and fallback to the usual ones.

Under proxy protocol, you have something called `    if (ngx_http_get_forwarded_addr(r, &addr, xfwd, value, rlcf->from,
                                    rlcf->recursive)
        != NGX_DECLINED)` which will be in charge of figuring out if that original IP was allowed to provide for a client IP with respect to `real_ip_recursive on/off` too.

So I think it's fine, except we need to improve the documentation on how to do all of this securely and why this is also important to firewall your flows to the PROXY-enabled listener if you are worried. :)

Reply:
(I have a small test with `curl --haproxy-protocol` that I'm trying to make work to spoof some IP and show that my theory is right.)

Reply:
(this is BTW further reinforced by guides like https://www.mythic-beasts.com/support/topics/proxy)

Reply:
So… I dropped the patch as it is working without it :). I misunderstood everything about how NGINX worked **and** I was not able to reproduce my initial issues without the patch.

I tried to cook up some attack tests but `curl` is not really good for this usecase, I reached the author to ask if he was okay with more flexibility on this which would open up to attack testing, but this is not ready for now.

I put it as a FIXME and will revisit it in the next months, in the meantime, I will just do basic sanity checks by hand on deployed servers.